### PR TITLE
Generate a unique identifier for autorewrite hints.

### LINKED ISSUE #16152: Deduplicate rewrite hints imported multiple times
```coq
Require Import Coq.Arith.PeanoNat.

Module MyHints.
  #[export] Hint Rewrite Nat.sub_diag : mydb.
End MyHints.

Fixpoint gen(n: nat) :=
  match n with
  | O => O
  | S m => n + gen m + n + gen m
  end.

Import MyHints.
Print Rewrite HintDb mydb.
(*
Database mydb
rewrite -> Nat.sub_diag of type forall n : nat, n - n = 0
*)

Goal gen 13 = 0.
Proof.
  unfold gen.
  Time autorewrite with mydb. (* 0.559 secs *)
  Time try rewrite_db mydb. (* 0.377 secs *)
Abort.

Import MyHints.
Print Rewrite HintDb mydb.
(*
Database mydb
rewrite -> Nat.sub_diag of type forall n : nat, n - n = 0
rewrite -> Nat.sub_diag of type forall n : nat, n - n = 0
*)

Goal gen 13 = 0.
Proof.
  unfold gen.
  Time autorewrite with mydb. (* 1.303 secs *)
  Time try rewrite_db mydb. (* 0.374 secs *)
Abort.

Import MyHints. Import MyHints. Import MyHints. Import MyHints.
Import MyHints. Import MyHints. Import MyHints. Import MyHints.

Goal gen 13 = 0.
Proof.
  unfold gen.
  Time autorewrite with mydb. (* 5.391 secs *)
  Time try rewrite_db mydb. (* 0.375 secs *)
Abort.
```

If I import the a module containing rewrite hints `n` times (which is likely to happen if one imports several files that do `Require Export`), the command `Print Rewrite HintDb mydb.` prints each hint `n` times, and `autorewrite` becomes `n` times slower.
Desired behavior: Each rewrite hint should only be printed once, and `autorewrite` should try each hint only once, no matter how many times it has been imported.

Coq version: 8.15-rc1

### PR DESCRIPTION
**Pull request by ppedrot**
Fixes #16152: Deduplicate rewrite hints imported multiple times.

### CODE REVIEW

**Comment by ppedrot** [comment]
@coqbot bench

**Comment by ppedrot** [comment]
@coqbot minimize

**Comment by JasonGross** [comment]
Presumably this is hitting the difference between `autorewrite` and `repeat autorewrite`, or is hitting even more subtle order sensitivity.

Also, you want @coqbot CI minimize

**Comment by JasonGross** [comment]
@ppedrot Do you have any idea what's going on with metacoq here?  The minimizer is saying that making some subset of the following changes is enough to make the regression go away:
- fully qualifying the `Require`s
- moving the file so that its fully qualified name is `PCUICInstConv` rather than `MetaCoq.PCUIC.Conversion.PCUICInstConv`

Maybe I should make it mimic the full binding?

**Comment by JasonGross** [comment]
Maybe before you try to merge this change, you should try out a PR that makes `autorewrite` and `repeat autorewrite` behave identically, and make `autorewrite` depend only on the ordering of the initial occurrence of a hint?

File: tactics/autorewrite.ml [hunk 1]
```diff
@@ -38,11 +38,11 @@ let subst_hint subst hint =
 
 module HintIdent =
 struct
-  type t = int * rew_rule
+  type t = KerName.t * rew_rule
 
-  let compare (i, t) (j, t') = i - j
+  let compare (kn, t) (kn', t') = KerName.compare kn kn'
```
Comment (JasonGross) [inline tactics/autorewrite.ml:43]: Is comparison ever used for sorting?

File: tactics/autorewrite.ml [hunk 2]
```diff
@@ -38,11 +38,11 @@ let subst_hint subst hint =
...
 
-  let subst s (i,t) = (i,subst_hint s t)
+  let subst s (kn, t) = (subst_kn s kn, subst_hint s t)
 
   let constr_of (i,t) = t.rew_pat
```
Comment (JasonGross) [inline tactics/autorewrite.ml:47]:
```suggestion
  let constr_of (_,t) = t.rew_pat
```
?

File: tactics/autorewrite.ml (hunk 1 above)
Comment (ppedrot) [reply tactics/autorewrite.ml:43]: It probably affects the order in which rewrite hints are tried, if several match. That's probably the root of the issue.

**Comment by ppedrot** [comment]
The CompCert failure is definitely due to a change of order of resolution. The final goal is the same but it changes the order of side condition subgoals.

**Comment by JasonGross** [comment]
> > Do you have any idea what's going on with metacoq here?
> 
> @JasonGross I'd bet that autorewrite is very order sensitive, and the hard-to-track transitivity of the requires affects the exact insertion order.

I suspect this is actually just a difference in ordering based on the changed kernel names of the lemmas in the current file; I'm not actually changing any of the `Requires`, just, e.g., changing `Require Logic` to `Require Coq.Init.Logic`, etc.

Can you pull some trick where you order by a pair of the old integer and the kernel name, or something?  I think autorewrite should either consistently prefer lemmas added earlier, or it should consistently prefer lemmas added later.

**Comment by ppedrot** [comment]
It's a bit hard to fix without some rework of the dnet structure, so that it behaves the same way as auto hint databases. I guess we can't spare a preliminary cleanup.

**Comment by JasonGross** [comment]
The failure to inline Crypto.Util.ZUtil.Div in f-c-l seems to reveal an independent bug in Coq, since the error message is "Error: Anomaly "Stm.join: tip not cached"".  Let's see if I can get the minimizer to minimize this.
@coqbot resume ci minimize ci-fiat_crypto_legacy
```
Require Coq.ZArith.ZArith.
Require Coq.micromega.Psatz.
Require Coq.Classes.Morphisms.
Require Coq.Classes.Morphisms_Prop.
Require Coq.micromega.Lia.
Require Coq.ZArith.Znumtheory.
Require Crypto.Util.GlobalSettings.
Require Crypto.Util.FixCoqMistakes.
Require Crypto.Util.ZUtil.Tactics.CompareToSgn.
Require Crypto.Util.ZUtil.Hints.Core.
Require Crypto.Util.ZUtil.Div.Bootstrap.
Require Crypto.Util.ZUtil.Modulo.Bootstrap.
Require Crypto.Util.ZUtil.Tactics.DivModToQuotRem.
Require Coq.Bool.Bool.
Require Crypto.Util.Bool.
Require Crypto.Util.ZUtil.Tactics.LtbToLt.
Require Coq.Classes.RelationClasses.
Require Coq.Lists.List.
Require Crypto.Util.ZUtil.Le.
Require Crypto.Util.ZUtil.Hints.ZArith.
Require Crypto.Util.ZUtil.Hints.PullPush.
Require Crypto.Util.ZUtil.Hints.Ztestbit.
Require Crypto.Util.ZUtil.ZSimplify.Core.
Require Crypto.Util.ZUtil.Hints.
Require Crypto.Util.Tactics.Head.
Require Crypto.Util.Tactics.BreakMatch.
Require Crypto.Util.ZUtil.Div.
Require Coq.ZArith.Zpow_facts.
Require Crypto.Util.ZUtil.Tactics.ReplaceNegWithPos.
Require Crypto.Util.ZUtil.Modulo.PullPush.
Require Crypto.Util.ZUtil.Tactics.PullPush.Modulo.
Require Crypto.Util.Tactics.DestructHyps.
Require Crypto.Util.Tactics.DestructHead.
Require Crypto.Util.ZUtil.Modulo.
Require Crypto.Util.ZUtil.Pow.
Require Crypto.Util.ZUtil.Tactics.PrimeBound.
Require Crypto.Util.ZUtil.Tactics.ZeroBounds.
Require Crypto.Util.ZUtil.Tactics.SimplifyFractionsLe.
Require Crypto.Util.ZUtil.ZSimplify.Autogenerated.
Require Crypto.Util.ZUtil.ZSimplify.Simple.
Require Crypto.Util.ZUtil.ZSimplify.

Declare ML Module "coq-core.plugins.ltac".
Module Export AdmitTactic.
Module Import LocalFalse.
Inductive False : Prop := .
End LocalFalse.
Axiom proof_admitted : False.
Global Set Default Proof Mode "Classic".
Tactic Notation "admit" := abstract case proof_admitted.
End AdmitTactic.
Import Coq.ZArith.ZArith Coq.micromega.Psatz Coq.Classes.Morphisms Coq.Classes.Morphisms_Prop.
Import Crypto.Util.ZUtil.Div.
Import Crypto.Util.ZUtil.Modulo.
Import Crypto.Util.ZUtil.Pow.
Import Crypto.Util.ZUtil.Tactics.ZeroBounds.
Import Crypto.Util.ZUtil.Tactics.SimplifyFractionsLe.
Import Crypto.Util.ZUtil.Tactics.LtbToLt.
Import Crypto.Util.ZUtil.Tactics.DivModToQuotRem.
Import Crypto.Util.ZUtil.ZSimplify.
Import Crypto.Util.ZUtil.Hints.ZArith.
Import Crypto.Util.Tactics.BreakMatch.

Local Open Scope Z_scope.

Section barrett.
  Context (n a : Z)
          (n_reasonable : n <> 0).






  Section general_idea.





    Local Notation "⌊am⌋" := (a / n) (only parsing).

    Theorem naive_barrett_reduction_correct
      : a mod n = a - ⌊am⌋ * n.
    Proof using n_reasonable.
      apply Zmod_eq_full; assumption.
    Qed.
  End general_idea.


  Section barrett_algorithm.




    Context (b : Z)
            (base_good : 0 < b)
            (k : Z)
            (k_good : n < b ^ k)
            (m : Z)
            (m_good : m = b^(2*k) / n)
            (offset : Z)
            (offset_nonneg : 0 <= offset).

    Context (n_pos : 0 < n)
            (a_nonneg : 0 <= a).

    Context (k_big_enough : offset <= k)
            (a_small : a < b^(2*k))

            (n_large : a mod b^(k-offset) <= n).



    Let q := (m * (a / b^(k-offset))) / b^(k+offset).
    Let r := a - q * n.

    Theorem barrett_reduction_equivalent
      : r mod n = a mod n.
    Proof using m_good offset.
      subst r q m.
      rewrite <- !Z.add_opp_r, !Zopp_mult_distr_l, !Z_mod_plus_full by assumption.
      reflexivity.
    Qed.

    Lemma qn_small
      : q * n <= a.
    Proof using a_nonneg a_small base_good k_big_enough m_good n_pos n_reasonable offset_nonneg.
      subst q r m.
      assert (0 < b^(k-offset)).
Z.zero_bounds.
      assert (0 < b^(k+offset)) by Z.zero_bounds.
      assert (0 < b^(2 * k)) by Z.zero_bounds.
      Z.simplify_fractions_le.
      autorewrite with pull_Zpow pull_Zdiv zsimplify; reflexivity.
    Qed.

    Lemma q_nice : { b : bool * bool | q = a / n + (if fst b then -1 else 0) + (if snd b then -1 else 0) }.
    Proof using a_nonneg a_small base_good k_big_enough m_good n_large n_pos n_reasonable offset_nonneg.
      assert (0 < b^(k+offset)) by Z.zero_bounds.
      assert (0 < b^(k-offset)) by Z.zero_bounds.
      assert (a / b^(k-offset) <= b^(2*k) / b^(k-offset)) by auto with zarith lia.
      assert (a / b^(k-offset) <= b^(k+offset)) by (autorewrite with pull_Zpow zsimplify in *; assumption).
      subst q r m.
      rewrite (Z.div_mul_diff_exact''' (b^(2*k)) n (a/b^(k-offset))) by auto with lia zero_bounds.
      rewrite (Z_div_mod_eq (b^(2*k) * _ / n) (b^(k+offset))) by lia.
      autorewrite with push_Zmul push_Zopp zsimplify zstrip_div zdiv_to_mod.
      rewrite Z.div_sub_mod_cond, !Z.div_sub_small by auto with zero_bounds zarith.
      eexists (_, _); reflexivity.
```

**Comment by JasonGross** [comment]
STM anomaly minimized to https://github.com/coq/coq/issues/16335

**Comment by JasonGross** [comment]
Okay, I think what's going on with f-c-l is: inlining Crypto.Util.ZUtil.Div changes kernel names in a way that results in a different ordering of autorewrite hints, making the error go away.  However, rather than just seeing a lack of error, instead https://github.com/coq/coq/issues/16335 shows up, because apparently master now raises anomalies when you leave a proof pending.

**Comment by JasonGross** [comment]
And hopefully I've fixed the minimizer to work with @coqbot ci minimize ci-metacoq

**Pull request closed without merging**